### ISSUE TITLE
nix eval .# is failing with eval error unless called with --raw or --json

### ISSUE THREAD

**Issue by andir** [NixOS/nix#5731]
**Describe the bug**

Unless called with `--raw` or `--json`, `nix eval` of a local flake doesn't do what is expected. I would expect it to eval all my outputs, checks, default packages etc... but not to end up somewhere in nixpkgs trying to access a meta attribute.

**Steps To Reproduce**
```console
$ mkdir testflake 
$ cd testflake 
$ nix flake init
$ nix eval .#     
warning: creating lock file '/home/andi/dev/testflake/flake.lock'
error: anonymous function at /nix/store/iq73m4fb8dc90l00fadq18wja61bcsc8-source/pkgs/build-support/fetchurl/boot.nix:5:1 called with unexpected argument 'meta'

       at /nix/store/iq73m4fb8dc90l00fadq18wja61bcsc8-source/pkgs/build-support/fetchzip/default.nix:22:2:

           21|
           22| (fetchurl (let
             |  ^
           23|   tmpFilename =
(use '--show-trace' to show detailed location information)
$ nix eval --raw .#
/nix/store/xcp9cav49dmsjbwdjlmkjxj10gkpx553-hello-2.10%                                                                                                                                                                     λ ~/dev/testflake/ car flake.nix            
$ cat flake.nix
{
  description = "A very basic flake";

  outputs = { self, nixpkgs }: {

    packages.x86_64-linux.hello = nixpkgs.legacyPackages.x86_64-linux.hello;

    defaultPackage.x86_64-linux = self.packages.x86_64-linux.hello;

  };
}
```

**Expected behavior**

Initially I was trying to find a way to have the equivalent behaviour of `nix-instantiate` with flakes but failed to find one. `nix eval` seemed like the closest thing to instantiating an expression.

Thinking about it more I'd expect `nix eval .#` to produce something like:
```
packages:
   * packageA: <Derivation /nix/store/.....drv or a list of out paths, or both?>
defaultPackage:
   * packageA: <Derivation /nix/store/.....drv or a list of out paths, or both?>
checks:
   * nixosVMTests.someTest: < ... same as above .... >
...
```

**`nix-env --version` output**
`nix-env (Nix) 2.5pre20211126_55275fc`

**Comment by cole-h** [comment]
Related nixpkgs issue: https://github.com/NixOS/nixpkgs/issues/97176

**Comment by regnat** [comment]
> Initially I was trying to find a way to have the equivalent behaviour of nix-instantiate with flakes but failed to find one. nix eval seemed like the closest thing to instantiating an expression.

You probably want `nix build --derivation`. `nix eval` is really about evaluating an expression, without any notion of what a `drv` file is (you can use `nix eval` if you really want to, but then you rather want `nix eval .#packageA.drvPath` or something like that)

> Thinking about it more I'd expect nix eval .# to produce something like:
>
> packages:
>    * packageA: <Derivation /nix/store/.....drv or a list of out paths, or both?>
> defaultPackage:
>    * packageA: <Derivation /nix/store/.....drv or a list of out paths, or both?>
> checks:
>    * nixosVMTests.someTest: < ... same as above .... >
...

`nix flake show` does kind-of that (though it doesn’t print the drv paths. But I’m not sure it would really make sense to do so in that case)

**Comment by dhess** [comment]
@andir I want what you want. Have you figured out how to do this with flakes?

**Comment by andir** [comment]
> @andir I want what you want. Have you figured out how to do this with flakes?

Nope. I just stopped using flakes. It doesn't help with developer UX.

**Comment by lilyball** [comment]
This sounds similar to my issue, which is that `nix eval nixpkgs#hello` (or any package) with do one of two things:

1. It will hang for as long as I'm willing to wait. I have not given it more than a minute or two, but that's more than enough time. I suspect it's recursively evaluating `stdenv`.
2. It will spit out an error. Earlier today `nix eval self#darwinConfigurations.DeerBook.pkgs.hello` was spitting out an error about the attribute `lib` not existing in the expression `${cc.lib}` within some llvm derivation, though right now it's hanging. `nix eval nixpkgs#hello` spits out an error about `Package ‘perl5.34.0-Alien-LibGumbo-0.05’` being marked as broken, which is remarkably bizarre (`nixpkgs` in this case is locked registry entry to a source path that is equivalent to `nixpkgs/81f05d871faf75d1456df6adec1d2118d787f65c`; using `nix eval nixpkgs/81f05d871faf75d1456df6adec1d2118d787f65c#hello` just hangs). I assume evaluation caches are somehow related to whether it hangs forever or returns, but I don't know the details.

My expectation here is that `nix eval some-drv` should just print something like `nix repl` does, such as `«derivation /nix/store/bdxy7dq6l9xxpl2k2jx4sckgn93rhi0v-hello-2.10.drv»`. Alternatively it could do like `nix-instantiate --eval` where you get a large dump of output that contains a lot of `<CODE>` entries, but that's really not useful at all. Maybe a middle ground is good where it shows the `nix repl` output plus all of the top-level attributes, but does not recurse.
